### PR TITLE
GCW-3311- sorting & filtering of goods on an order view

### PR DESCRIPTION
**Pull request by abulsayyad123**
### Ticket Link: 

https://jira.crossroads.org.hk/browse/GCW-3311

### What does this PR do?

- Add query for searching, filtering and sorting orders_packages for given order_id.

### CODE REVIEW

File: app/models/concerns/orders_package_search.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+  included do
+    scope :search, ->(options = {}) {
+      search_text = options[:search_text].downcase || ''
+      if search_text.present?
+        search_query = SEARCH_ATTRIBUTES.map { |f| "#{f} ILIKE :search_text" }.join(" OR ")
+        where(search_query, search_text: "%#{search_text}%").distinct
+      else
+        none
+      end
```
Comment (bharat619) [inline app/models/concerns/orders_package_search.rb:13]:
```suggestion
    scope :search, ->(options = {}) {
      search_text = options[:search_text].downcase || ''
      return nil unless search_text.present?
      
      search_query = SEARCH_ATTRIBUTES.map { |f| "#{f} ILIKE :search_text" }.join(" OR ")
       where(search_query, search_text: "%#{search_text}%").distinct
```

File: app/models/orders_package.rb [hunk 2]
```diff
@@ -29,13 +30,30 @@ module States
   scope :for_order, ->(order_id) { joins(:order).where(orders: { id: order_id }) }
   scope :not_cancellable, -> () { where("orders_packages.state = 'dispatched' OR dispatched_quantity > 0") }
   scope :cancellable, -> () { where("orders_packages.state = 'designated' AND dispatched_quantity = 0") }
+  scope :sorting, -> (options) { order(sort_orders_package(options)) }
+  scope :get_by_state, ->(states) { where("orders_packages.state IN (?)", states) }
```
Comment (bharat619) [inline app/models/orders_package.rb:34]:
```suggestion
  scope :by_state, ->(states) { where("orders_packages.state IN (?)", states) }
```

File: app/models/orders_package.rb [hunk 3]
```diff
@@ -29,13 +30,30 @@ module States
...
     includes([
       { package: [ :locations, {package_type: [:location]}, :images, :orders_packages] }
     ])
   }
 
+  def self.search_and_filter(options)
+    orders_packages = joins(package: [:package_type])
+    orders_packages = orders_packages.select("orders_packages.*, package_types.code, package_types.name_en, packages.inventory_number")
+    orders_packages = orders_packages.search(options) if options[:search_text]
+    orders_packages = orders_packages.get_by_state(options[:state_names]) if options[:state_names].any?
```
Comment (bharat619) [inline app/models/orders_package.rb:46]:
```suggestion
    orders_packages = orders_packages.by_state(options[:state_names]) if options[:state_names].any?
```

**Review by bharat619 — APPROVED** [review]
Optional changes suggested

File: app/models/concerns/orders_package_search.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
+module OrdersPackageSearch
+  extend ActiveSupport::Concern
+  SEARCH_ATTRIBUTES = ['package_types.code', 'package_types.name_en', 'packages.inventory_number'].freeze
+
+  included do
+    scope :search, ->(options = {}) {
+      search_text = options[:search_text].downcase || ''
+      return nil unless search_text.present?
```
Comment (swatijadhav) [inline app/models/concerns/orders_package_search.rb:8]:
```suggestion
      return [] unless search_text.present?
```

**Review by shreyas098 — APPROVED** [review]

**Review by swatijadhav — APPROVED** [review]

**Pull request merged**